pkg/netceptor/packetconn.go

```diff
@@ -42,35 +79,46 @@ func (s *Netceptor) ListenPacket(service string) (*PacketConn, error) {
 		return nil, fmt.Errorf("service %s is already listening", service)
 	}
 	_ = s.addNameHash(service)
-	pc := &PacketConn{
-		s:            s,
-		localService: service,
-		recvChan:     make(chan *MessageData),
-		advertise:    false,
-		adTags:       nil,
-		connType:     ConnTypeDatagram,
-		hopsToLive:   s.maxForwardingHops,
-	}
-	pc.startUnreachable()
-	s.listenerRegistry[service] = pc
+	pc := NewPacketConn(s, service, ConnTypeDatagram)
 
 	return pc, nil
 }
 
 // ListenPacketAndAdvertise returns a datagram listener, and also broadcasts service
 // advertisements to the Receptor network as long as the listener remains open.
-func (s *Netceptor) ListenPacketAndAdvertise(service string, tags map[string]string) (*PacketConn, error) {
-	pc, err := s.ListenPacket(service)
-	if err != nil {
-		return nil, err
+func (s *Netceptor) ListenPacketAndAdvertise(service string, tags map[string]string) (PacketConner, error) {
+	if len(service) > 8 {
+		return nil, fmt.Errorf("service name %s too long", service)
 	}
-	pc.advertise = true
-	pc.adTags = tags
+	if service == "" {
+		service = s.getEphemeralService()
+	}
+	s.listenerLock.Lock()
+	defer s.listenerLock.Unlock()
+	_, isReserved := s.reservedServices[service]
+	_, isListening := s.listenerRegistry[service]
+	if isReserved || isListening {
+		return nil, fmt.Errorf("service %s is already listening", service)
```

Comment:
Lets make this error a little more verbose so we can tell the difference between here and line 79.

Im open to ideas, here is my suggestion `""service %s is already listening and advertising", service"`